R/as-tibble.R

```diff
@@ -1,26 +1,77 @@
+#' multidimensional array. In this form each dimension of the cube is a variable
+#' and the cube values are expressed columns for each measure. This is useful
+#' both to better understand what each entry of a cube represents, and to work
+#' with the cube result using tidyverse tools.
+#'
+#' @param x a CrunchCube
+#' @param ... futher arguments passed on to `tibble::as_tibble()`
+#'
 #' @export
 #' @importFrom tibble as_tibble
+#' @importFrom crunch getDimTypes
+#' @importFrom dplyr bind_cols
+#' @importFrom purrr map2 map reduce
+#' @importFrom stringr str_extract
 as_tibble.CrunchCube <- function (x, ...) {
     ## TODO: Consider using `dplyr::tbl_cube` class
-    ## TODO: NA behavior? Add way to get missings not missing. (This uses
-    ## as.array to remove extraneous dims, and doing so applies the '@useNA'
-    ## from the cube)
-    ## TODO: deal with naming for array (and NUMR) variable dims (var.categories?)
+    
+    dnames <- dimnames(x@arrays$.unweighted_counts)
+    measures <- names(x@arrays)
 
-    ## 1) get the measures
-    ## TODO: better access methods for secondary measures in Cube
-    ## TODO: if weighted query, include ".unweighted_counts"
-    measure_names <- setdiff(names(x@arrays), ".unweighted_counts")
-    vals <- sapply(measure_names, function (m) {
-        as.vector(crunch:::cubeToArray(x, m))
-    }, simplify=FALSE)
+    # Cubes can have multiple measures, which are represented as their own column
+    # in the tibble.
+    measure_vals <- sapply(measures, function(y) {
+        as.vector(x@arrays[[y]])
+    }, simplify = FALSE)
+    
+    names(measure_vals)[names(measure_vals) == ".unweighted_counts"] <- "row_count"
 
-    ## 2) If there are dimnames, expand.grid and c(bind) them
-    dnames <- dimnames(as.array(x))
-    ## NULL if scalar value, which means no "group_by"
+    # If there are dimnames, expand.grid and bind them. We also change the names
+    # of the two array dimensions to avoid duplicated variable names in  the
+    # tibble.
     if (!is.null(dnames)) {
-        dims <- do.call(expand.grid, dnames)
-        vals <- c(dims, vals)
+        types <- getDimTypes(x)
+
+        # Change MR selection vars to T/F/NA
+        is_selected <- types == "mr_selections"
+        dnames <- map2(dnames, is_selected, ~{
+            if (.y) {
+                return(c(TRUE, FALSE, NA))
+            } else {
+                return (.x)
+            }
+        })
+        suffixes <- str_extract(types, "_.*$")
+        is_array_var <- !is.na(suffixes)
+        names(dnames)[is_array_var] <- paste0(names(dnames)[is_array_var], suffixes[is_array_var])
+        out <- do.call(expand.grid, dnames)
+        names(out) <- add_duplicate_suffix(names(out))
+        
+        # Identify which elements of cube represent missing values
+        out$is_missing <- x@dims %>%
+            map("missing") %>%
+            expand.grid() %>%
+            reduce(`|`)
```

Comment:
Am I reading this correctly that you're stringing together the missing elements with `|` between them? Any reason not to keep them as a vector?

Reply:
I might be doing a bit much with the pipeline here. The reduce is reducing between the logical vectors in the data.frame, so what you get out is a single logical vector representing whether there are any TRUEs in the underlying dimension. I think this is a safe way of doing it but let me know if you can think of a case I'm missing here. 

```
x@dims %>% (list)
            map("missing") %>% (list of short logical vectors)
            expand.grid() %>% (data frame each column is a dimension, row is the unique combination of dimensions)
            reduce(`|`) ( df$v1 | df$v2 | dfv3 .... | df$vn)
```

Initially I kept each dimension's missingness as its own column, but in working with the output I found that I only ever wanted to know whether any dimension was missing for a given entry. The thought was that if any dimension is missing then the entry is also missing right?